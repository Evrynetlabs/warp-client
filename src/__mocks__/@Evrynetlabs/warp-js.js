export const spyToEvrynet = jest.fn()
export const spyToStellar = jest.fn()
export const spyGetWhitelistAssets = jest.fn()
export const spyGetCode = jest.fn()
export const spyGetAccountBalance = jest.fn()
export const spyGetPublickeyFromPrivateKey = jest.fn()
export const spyGetTrustlines = jest.fn()
export const spyIsListening = jest.fn()

const mock = jest.fn().mockImplementation(() => {
  return {
    toEvrynet: spyToEvrynet,
    toStellar: spyToStellar,
    client: {
      evry: {
        getWhitelistAssets: spyGetWhitelistAssets,
        getBalance: spyGetAccountBalance,
        getPublickeyFromPrivateKey: spyGetPublickeyFromPrivateKey,
      },
      stellar: {
        getBalance: spyGetAccountBalance,
        getPublickeyFromPrivateKey: spyGetPublickeyFromPrivateKey,
        getTrustlines: spyGetTrustlines,
      },
      net: {
        isListening: spyIsListening,
      },
    },
    utils: {
      getEvryAsset: jest.fn().mockReturnValue({
        getCode: spyGetCode,
      }),
      WhitelistedAsset: {
        NATIVE_ASSET: 1,
      },
    },
  }
})
export const WarpConfig = jest.fn().mockImplementation(() => {
  return {
    evrynet: {
      gasLimit: 2000000,
      atomicEvryDecimalUnit: 18,
      atomicStellarDecimalUnit: 7,
    },
  }
})

export const warpConfigInstance = WarpConfig()

export const initWarpConfig = jest.fn()
export default mock
