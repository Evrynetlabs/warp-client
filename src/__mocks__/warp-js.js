// __mocks__/warp-js.js

// Import this named export into your test file
export const spyToEvrynet = jest.fn()
const mock = jest.fn().mockImplementation(() => {
  return {
    toEvrynet: spyToEvrynet,
    client: jest.fn().mockReturnValue({
      evry: jest.fn().mockReturnValue({
        getWhitelistAssets: jest.fn(),
      }),
    }),
  }
  // Now we can track calls to playSoundFile
})
export default mock
