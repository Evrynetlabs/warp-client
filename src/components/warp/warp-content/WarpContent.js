import React, { Component } from 'react'
import Card from 'react-bootstrap/Card'
import classNames from 'classnames'
import { Container, Row, Col, Button, Form } from 'react-bootstrap'
import 'Components/warp/warp-content/warpContent.scss'
import { numberToCurrencyString, currencyToNumberString } from '@/utils/format'
import PropTypes from 'prop-types'
import Warp from 'warp-js'
import map from 'lodash/map'
import isEmpty from 'lodash/isEmpty'
import find from 'lodash/find'

export default class WarpContent extends Component {
  constructor(props) {
    super(props)
    this.warp = new Warp()
    this.initialState = {
      styles: this._initStyles(),
      amount: '',
      asset: this.warp.utils.getEvryAsset().getCode(),
      role: {
        source: 'source',
        destination: 'destination',
      },
      sourceAccount: '',
      destinationAccount: '',
      transferFunc: props.toEvry,
    }
    this.state = {
      ...this.initialState,
    }
  }

  _initStyles() {
    const stylesMain = classNames({
      [this.constructor.name]: true,
    })
    const stylesFooter = `${stylesMain}__footer`
    const stylesForm = `${stylesMain}__form`
    const stylesContent = `${stylesForm}__content`
    const stylesContentAccountInput = `${stylesContent}__input`
    const stylesContentAccountInputSrc = `${stylesContentAccountInput}__src`
    const stylesContentAccountInputDest = `${stylesContentAccountInput}__dest`
    const stylesContentAmountSelection = `${stylesContent}__amount`
    const stylesFooterButton = `${stylesFooter}__btn`
    return {
      main: stylesMain,
      content: stylesContent,
      footer: stylesFooter,
      footerBtn: stylesFooterButton,
      amountSelection: stylesContentAmountSelection,
      accountInput: stylesContentAccountInput,
      accountInputSrc: stylesContentAccountInputSrc,
      accountInputDest: stylesContentAccountInputDest,
      form: stylesForm,
    }
  }

  _formatNumber(amount) {
    let decimal
    if (!isEmpty(this.props.whitelistedAssets.state)) {
      const whitelistedAsset = find(this.props.whitelistedAssets.state, {
        code: this.state.asset,
      })
      decimal = whitelistedAsset ? whitelistedAsset.decimal : whitelistedAsset
    }
    this.setState({
      amount: numberToCurrencyString(
        Number(currencyToNumberString(amount)),
        decimal,
      ),
    })
  }

  _saveAmount(e) {
    this.setState({
      amount: e.target.value,
    })
  }

  _saveAsset(e) {
    this.setState({
      asset: e.target.value,
    })
  }

  _saveTransactionAccount(e, role = '') {
    this.setState({
      ...(this.state.role.source === role && { sourceAccount: e.target.value }),
      ...(this.state.role.destination === role && {
        destinationAccount: e.target.value,
      }),
    })
  }

  async componentDidMount() {
    await this.props.getWhitelistAssets()
  }

  async _handleSubmit(e) {
    e.preventDefault()
    let asset
    switch (this.state.asset) {
      case 'EVRY': {
        asset = this.warp.utils.getEvryAsset()
        break
      }
      case 'XLM': {
        asset = this.warp.utils.getLumensAsset()
        break
      }
      default:
        return null
    }
    await this.state.transferFunc({
      asset,
      amount: currencyToNumberString(this.state.amount),
      src: this.state.sourceAccount,
      dest: this.state.destinationAccount,
    })
  }

  _listWhitelistedAssetsOptions() {
    if (
      this.props.whitelistedAssets.loading ||
      isEmpty(this.props.whitelistedAssets.state)
    )
      return
    return map(this.props.whitelistedAssets.state, (ech) => {
      return (
        <option key={ech.getCode()} value={ech.getCode()}>
          {ech.getCode()}
        </option>
      )
    })
  }

  __updateTransferFunction(prevProps, prevState) {
    if (prevProps.isToEvry === this.props.isToEvry) return
    if (this.props.isToEvry) {
      this.setState({
        transferFunc: this.props.toEvry,
        sourceAccount: prevState.destinationAccount,
        destinationAccount: prevState.sourceAccount,
      })
      return
    }
    this.setState({
      transferFunc: this.props.toStellar,
      sourceAccount: prevState.destinationAccount,
      destinationAccount: prevState.sourceAccount,
    })
  }

  componentDidUpdate(prevProps, prevState) {
    this.__updateTransferFunction(prevProps, prevState)
  }

  render() {
    return (
      <Form
        className={this.state.styles.form}
        onSubmit={async (e) => {
          await this._handleSubmit(e)
        }}
      >
        <Card.Body className={this.state.styles.content}>
          <Container fluid>
            <Row>
              <Col>
                <Form.Group controlId="sourceAccountNumber">
                  <Form.Label>
                    <span>From:</span>
                  </Form.Label>
                  <Form.Control
                    className={classNames({
                      [this.state.styles.accountInputSrc]: true,
                      [this.state.styles.accountInput]: true,
                    })}
                    type="text"
                    placeholder="Account Number"
                    value={this.state.sourceAccount}
                    onChange={(e) =>
                      this._saveTransactionAccount(e, this.state.role.source)
                    }
                  />
                </Form.Group>
              </Col>
              <Col>
                <Form.Group controlId="destinationAccountNumber">
                  <Form.Label>
                    <span>To:</span>
                  </Form.Label>
                  <Form.Control
                    className={classNames({
                      [this.state.styles.accountInputDest]: true,
                      [this.state.styles.accountInput]: true,
                    })}
                    type="text"
                    placeholder="Account Number"
                    value={this.state.destinationAccount}
                    onChange={(e) =>
                      this._saveTransactionAccount(
                        e,
                        this.state.role.destination,
                      )
                    }
                  />
                </Form.Group>
              </Col>
            </Row>
            <Row className={this.state.styles.amountSelection}>
              <Col>
                <Form.Group controlId="assetAmount">
                  <Form.Control
                    type="text"
                    onChange={(e) => this._saveAmount(e)}
                    onBlur={() => this._formatNumber(this.state.amount)}
                    placeholder="0.00"
                    value={this.state.amount}
                  ></Form.Control>
                </Form.Group>
              </Col>
              <Col>
                <Form.Group controlId="assetSelection">
                  <Form.Control
                    value={this.state.asset}
                    as="select"
                    onChange={(e) => this._saveAsset(e)}
                  >
                    {this._listWhitelistedAssetsOptions()}
                  </Form.Control>
                </Form.Group>
              </Col>
            </Row>
          </Container>
        </Card.Body>
        <Card.Footer className={this.state.styles.footer}>
          <Container fluid>
            <Row>
              <Col className={this.state.styles.footerBtn}>
                <Button type="submit" variant="dark">
                  Transfer
                </Button>
              </Col>
            </Row>
          </Container>
        </Card.Footer>
      </Form>
    )
  }
}

WarpContent.propTypes = {
  txHashes: PropTypes.shape({
    state: PropTypes.shape({
      stellar: PropTypes.string,
      evry: PropTypes.string,
    }),
    loading: PropTypes.bool,
    error: PropTypes.object,
  }),
  whitelistedAssets: PropTypes.shape({
    state: PropTypes.array,
    loading: PropTypes.bool,
    error: PropTypes.object,
  }),
  toEvry: PropTypes.func.isRequired,
  toStellar: PropTypes.func.isRequired,
  isToEvry: PropTypes.bool.isRequired,
  getWhitelistAssets: PropTypes.func.isRequired,
}
