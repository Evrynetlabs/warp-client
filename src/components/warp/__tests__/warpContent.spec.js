import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
jest.mock('warp-js')
import { spyGetCode } from 'warp-js'
import WarpContent from 'Components/warp/warp-content/WarpContent'
Enzyme.configure({ adapter: new Adapter() })
import { Form } from 'react-bootstrap'

describe('WarpContent', () => {
  const mock = {
    txHashes: {
      state: {
        stellar: 'foo',
        evry: 'bar',
      },
      loading: false,
      error: null,
    },
    toStellar: jest.fn(),
    toEvrynet: jest.fn(),
    whitelistedAssets: {
      state: [],
      loading: false,
      error: null,
    },
    accountBalance: {
      state: null,
      loading: false,
      error: null,
    },
    getWhitelistAssets: jest.fn(),
    isToEvrynet: true,
    getAccountBalance: jest.fn(),
  }
  spyGetCode.mockReturnValue('EVRY')
  let component

  beforeEach(() => {
    component = shallow(
      <WarpContent
        txHashes={mock.txHashes}
        toEvrynet={mock.toEvrynet}
        toStellar={mock.toStellar}
        whitelistedAssets={mock.whitelistedAssets}
        getWhitelistAssets={mock.getWhitelistAssets}
        isToEvrynet={mock.isToEvrynet}
        getAccountBalance={mock.getAccountBalance}
      ></WarpContent>,
    )
  })

  it('should render correspondingly', () => {
    expect(component).toMatchSnapshot()
  })

  describe('When input an account souce', () => {
    const expected = 'foo'
    const mockEvent = {
      target: {
        value: expected,
        name: 'sourceAccount',
      },
    }
    it('should save a source account state correspondingly when on changing input', () => {
      component
        .find('.WarpContent__form__content__input__src')
        .simulate('change', mockEvent)
      expect(component.state().formControls.sourceAccount.value).toEqual(
        expected,
      )
    })
    it('should get an error "Invalid Stellar secret key format."', () => {
      let mockEvent = {
        target: {
          value: 'foo',
          name: 'sourceAccount',
        },
      }

      component
        .find('.WarpContent__form__content__input__src')
        .simulate('change', mockEvent)

      expect(component.state().formControls.sourceAccount.valid).toEqual(false)

      expect(component.state().formControls.sourceAccount.errorMessage).toEqual(
        'Invalid Stellar secret key format.',
      )
    })

    it('should get an error "Stellar secret key is required."', () => {
      let mockEvent = {
        target: {
          value: '',
          name: 'sourceAccount',
        },
      }

      component
        .find('.WarpContent__form__content__input__src')
        .simulate('change', mockEvent)

      expect(component.state().formControls.sourceAccount.valid).toEqual(false)

      expect(component.state().formControls.sourceAccount.errorMessage).toEqual(
        'Stellar secret key is required.',
      )
    })

    it('should not get any errors', () => {
      let mockEvent = {
        target: {
          value: 'SCIMPYI2AZKATTOQOOE5LFHRHLIDSCJP2CRBHQ6ZKHVTASCXNOQQNMKE',
          name: 'sourceAccount',
        },
      }

      component
        .find('.WarpContent__form__content__input__src')
        .simulate('change', mockEvent)

      expect(component.state().formControls.sourceAccount.valid).toEqual(true)

      expect(component.state().formControls.sourceAccount.errorMessage).toEqual(
        null,
      )
    })
  })

  describe('When input an account destination', () => {
    const expected = 'foo'
    const mockEvent = {
      target: {
        value: expected,
        name: 'destinationAccount',
      },
    }
    it('should save a source account state correspondingly when on chaning input', () => {
      component
        .find('.WarpContent__form__content__input__dest')
        .simulate('change', mockEvent)
      expect(component.state().formControls.destinationAccount.value).toEqual(
        expected,
      )
    })
    it('should get an error "Invalid Evrynet secret key format."', () => {
      let mockEvent = {
        target: {
          value: 'foo',
          name: 'destinationAccount',
        },
      }

      component
        .find('.WarpContent__form__content__input__src')
        .simulate('change', mockEvent)

      expect(component.state().formControls.destinationAccount.valid).toEqual(
        false,
      )

      expect(
        component.state().formControls.destinationAccount.errorMessage,
      ).toEqual('Invalid Evrynet secret key format.')
    })

    it('should get an error "Evrynet secret key is required."', () => {
      let mockEvent = {
        target: {
          value: '',
          name: 'destinationAccount',
        },
      }

      component
        .find('.WarpContent__form__content__input__src')
        .simulate('change', mockEvent)

      expect(component.state().formControls.destinationAccount.valid).toEqual(
        false,
      )

      expect(
        component.state().formControls.destinationAccount.errorMessage,
      ).toEqual('Evrynet secret key is required.')
    })

    it('should not get any errors', () => {
      let mockEvent = {
        target: {
          value:
            'B9CD2CCDB7CCE1EE6A2C9C21A3A7F97506666D2B02946091F2476762FA1F1039',
          name: 'destinationAccount',
        },
      }

      component
        .find('.WarpContent__form__content__input__src')
        .simulate('change', mockEvent)

      expect(component.state().formControls.destinationAccount.valid).toEqual(
        true,
      )

      expect(
        component.state().formControls.destinationAccount.errorMessage,
      ).toEqual(null)
    })
  })

  describe('When input an amount', () => {
    const inputs = [
      ['0.112', '0.112'],
      ['1.0', '1.0'],
      ['1.01', '1.01'],
      ['1.012', '1.012'],
      ['1.112', '1.112'],
      ['10.112', '10.112'],
      ['1000.112', '1000.112'],
      ['1.0', '1.0'],
      ['0001.0', '1.0'],
      ['-1', '-1'],
      ['1.', '1'],
      ['001.', '1'],
    ]
    test.each(inputs)(
      'should create a correct currency string',
      (amount, expected) => {
        const mockEvent = {
          target: {
            value: amount,
            name: 'amount',
          },
        }
        const input = component.find(
          '.WarpContent__form__content__amount__input',
        )
        const mockWhitelistedAsset = {
          getCode: jest.fn().mockReturnValue('EVRY'),
          decimal: 18,
          getDecimal: jest.fn().mockReturnValue(18),
        }
        component.setProps({
          whitelistedAssets: {
            state: [mockWhitelistedAsset],
          },
        })
        input.simulate('change', mockEvent)
        expect(component.state().formControls.amount.value).toEqual(amount)
        input.simulate('blur', mockEvent)
        expect(component.state().formControls.amount.value).toEqual(expected)
      },
    )

    describe('When input an invalid amount', () => {
      describe('When amount is zero', () => {
        it('should save an invalid state with a desired error message', () => {
          const mockEvent = {
            target: {
              value: '0',
              name: 'amount',
            },
          }
          const input = component.find(
            '.WarpContent__form__content__amount__input',
          )
          input.simulate('change', mockEvent)
          expect(component.state().formControls.amount.valid).toEqual(false)
          expect(component.state().formControls.amount.errorMessage).toEqual(
            'Amount must be greater than zero.',
          )
          expect(component).toMatchSnapshot()
        })
      })

      describe('When amount is less than zero', () => {
        test.each(['-1', '-0.01', '-1.2222', '0', '0.000'])(
          'should save an invalid state with a desired error message',
          (ech) => {
            const mockEvent = {
              target: {
                value: ech,
                name: 'amount',
              },
            }
            const input = component.find(
              '.WarpContent__form__content__amount__input',
            )
            input.simulate('change', mockEvent)
            expect(component.state().formControls.amount.valid).toEqual(false)
            expect(component.state().formControls.amount.errorMessage).toEqual(
              'Amount must be greater than zero.',
            )
            expect(component).toMatchSnapshot()
          },
        )
      })

      describe('When amount precision exceeds limit decimal', () => {
        test.each([['1.000', 2], ['1.00000000', 7]])(
          'should save an invalid state with a desired error message',
          (ech, decimal) => {
            const mockEvent = {
              target: {
                value: ech,
                name: 'amount',
              },
            }
            const mockWhitelistedAsset = {
              getCode: jest.fn().mockReturnValue('EVRY'),
              decimal,
              getDecimal: jest.fn().mockReturnValue(decimal),
            }
            component.setProps({
              whitelistedAssets: {
                state: [mockWhitelistedAsset],
              },
            })
            const input = component.find(
              '.WarpContent__form__content__amount__input',
            )
            input.simulate('change', mockEvent)
            expect(component.state().formControls.amount.valid).toEqual(false)
            expect(component.state().formControls.amount.errorMessage).toEqual(
              `Amount can only support a precision of ${mockWhitelistedAsset.decimal} decimals.`,
            )
            expect(component).toMatchSnapshot()
          },
        )
      })

      describe('When amount is empty', () => {
        it('should display an invalid feedback', () => {
          const mockEvent = {
            target: {
              value: '',
              name: 'amount',
            },
          }
          const mockWhitelistedAsset = {
            getCode: jest.fn().mockReturnValue('EVRY'),
            decimal: 2,
            getDecimal: jest.fn().mockReturnValue(2),
          }
          component.setProps({
            whitelistedAssets: {
              state: [mockWhitelistedAsset],
            },
          })
          const input = component.find(
            '.WarpContent__form__content__amount__input',
          )
          input.simulate('change', mockEvent)
          expect(component.state().formControls.amount.valid).toEqual(false)
          expect(component.state().formControls.amount.errorMessage).toEqual(
            'Amount is required.',
          )
          expect(component).toMatchSnapshot()
        })
      })

      describe('When amount is not a number', () => {
        test.each(['--1.000', '1.00--0000000', 'foo', 'bar', '-.', '0.--1'])(
          'should save an invalid state with a desired error message',
          (ech) => {
            const mockEvent = {
              target: {
                value: ech,
                name: 'amount',
              },
            }
            const mockWhitelistedAsset = {
              getCode: jest.fn().mockReturnValue('EVRY'),
              decimal: 2,
              getDecimal: jest.fn().mockReturnValue(2),
            }
            component.setProps({
              whitelistedAssets: {
                state: [mockWhitelistedAsset],
              },
            })
            const input = component.find(
              '.WarpContent__form__content__amount__input',
            )
            input.simulate('change', mockEvent)
            expect(component.state().formControls.amount.valid).toEqual(false)
            expect(component.state().formControls.amount.errorMessage).toEqual(
              'Amount must be a number.',
            )
            expect(component).toMatchSnapshot()
          },
        )
      })
    })
  })

  describe('When clicking a transfer button', () => {
    describe('When isToEvrynet is false', () => {
      test('transfer function should be toStellar', () => {
        component.setProps({
          isToEvrynet: false,
        })
        expect(component.state().transferFunc).toEqual(mock.toStellar)
        expect(component.state().transferFunc).not.toEqual(mock.toEvrynet)
      })
      describe('When source accountbalance is invalid', () => {
        it('should show an invalid feedback', async () => {
          component.setProps({
            isToEvrynet: false,
            accountBalance: {
              ...component.props().accountBalance,
              state: '0',
            },
            whitelistedAssets: {
              state: [
                {
                  getCode: jest.fn().mockReturnValue('EVRY'),
                  decimal: 1,
                  getDecimal: jest.fn().mockReturnValue(1),
                },
              ],
            },
          })
          let updatedFormControls = { ...component.state().formControls }
          updatedFormControls.amount.value = '1'
          component.setState({
            formControls: updatedFormControls,
          })
          await component.instance()._submitHandler('amount')
          expect(component.state().formControls.amount.valid).toBe(false)
          expect(component.state().formControls.amount.errorMessage).toBe(
            'Insufficient Amount',
          )
        })
        it('should match an invalid feedback snapshot', async () => {
          component.setProps({
            isToEvrynet: false,
            accountBalance: {
              ...component.props().accountBalance,
              state: '0',
            },
            whitelistedAssets: {
              state: [
                {
                  getCode: jest.fn().mockReturnValue('EVRY'),
                  decimal: 1,
                  getDecimal: jest.fn().mockReturnValue(1),
                },
              ],
            },
          })
          let updatedFormControls = { ...component.state().formControls }
          updatedFormControls.amount.value = '1'
          component.setState({
            formControls: updatedFormControls,
          })
          const form = component.find(Form)
          await form.props().onSubmit({
            preventDefault: () => {},
            target: {
              name: 'amount',
            },
          })
          expect(component).toMatchSnapshot()
        })
      })

      describe('When source accountbalance is valid', () => {
        it('should show a valid feedback', async () => {
          component.setProps({
            isToEvrynet: false,
            accountBalance: {
              ...component.props().accountBalance,
              state: '10',
            },
            whitelistedAssets: {
              state: [
                {
                  getCode: jest.fn().mockReturnValue('EVRY'),
                  decimal: 1,
                  getDecimal: jest.fn().mockReturnValue(1),
                },
              ],
            },
          })
          let updatedFormControls = { ...component.state().formControls }
          updatedFormControls.amount.value = '1'
          component.setState({
            formControls: updatedFormControls,
          })
          await component.instance()._submitHandler('amount')
          expect(component.state().formControls.amount.valid).toBe(true)
          expect(component.state().formControls.amount.errorMessage).toBe('')
        })
        it('should match a valid feedback snapshot', async () => {
          component.setProps({
            isToEvrynet: false,
            accountBalance: {
              ...component.props().accountBalance,
              state: '10',
            },
            whitelistedAssets: {
              state: [
                {
                  getCode: jest.fn().mockReturnValue('EVRY'),
                  decimal: 1,
                  getDecimal: jest.fn().mockReturnValue(1),
                },
              ],
            },
          })
          let updatedFormControls = { ...component.state().formControls }
          updatedFormControls.amount.value = '1'
          component.setState({
            formControls: updatedFormControls,
          })
          const form = component.find(Form)
          await form.props().onSubmit({
            preventDefault: () => {},
            target: {
              name: 'amount',
            },
          })
          expect(component).toMatchSnapshot()
        })
      })
    })

    describe('When isToEvrynet is true', () => {
      test('transfer function should be toEvrynet', () => {
        expect(component.state().transferFunc).toEqual(mock.toEvrynet)
        expect(component.state().transferFunc).not.toEqual(mock.toStellar)
      })
      describe('When source accountbalance is invalid', () => {
        it('should show an invalid feedback', async () => {
          component.setProps({
            isToEvrynet: true,
            accountBalance: {
              ...component.props().accountBalance,
              state: '0',
            },
            whitelistedAssets: {
              state: [
                {
                  getCode: jest.fn().mockReturnValue('EVRY'),
                  decimal: 1,
                  getDecimal: jest.fn().mockReturnValue(1),
                },
              ],
            },
          })
          let updatedFormControls = { ...component.state().formControls }
          updatedFormControls.amount.value = '1'
          component.setState({
            formControls: updatedFormControls,
          })
          await component.instance()._submitHandler('amount')
          expect(component.state().formControls.amount.valid).toBe(false)
        })
        it('should match an invalid feedback snapshot', async () => {
          component.setProps({
            isToEvrynet: true,
            accountBalance: {
              ...component.props().accountBalance,
              state: '0',
            },
            whitelistedAssets: {
              state: [
                {
                  getCode: jest.fn().mockReturnValue('EVRY'),
                  decimal: 1,
                  getDecimal: jest.fn().mockReturnValue(1),
                },
              ],
            },
          })
          let updatedFormControls = { ...component.state().formControls }
          updatedFormControls.amount.value = '1'
          component.setState({
            formControls: updatedFormControls,
          })
          const form = component.find(Form)
          await form.props().onSubmit({
            preventDefault: () => {},
            target: {
              name: 'amount',
            },
          })
          expect(component).toMatchSnapshot()
        })
      })

      describe('When source accountbalance is valid', () => {
        it('should show a valid feedback', async () => {
          component.setProps({
            isToEvrynet: true,
            accountBalance: {
              ...component.props().accountBalance,
              state: '10000000',
            },
            whitelistedAssets: {
              state: [
                {
                  getCode: jest.fn().mockReturnValue('EVRY'),
                  decimal: 1,
                  getDecimal: jest.fn().mockReturnValue(1),
                },
              ],
            },
          })
          let updatedFormControls = { ...component.state().formControls }
          updatedFormControls.amount.value = '1'
          component.setState({
            formControls: updatedFormControls,
          })
          await component.instance()._submitHandler('amount')
          expect(component.state().formControls.amount.valid).toBe(true)
          expect(component.state().formControls.amount.errorMessage).toBe('')
        })
        it('should match a valid feedback snapshot', async () => {
          component.setProps({
            isToEvrynet: true,
            accountBalance: {
              ...component.props().accountBalance,
              state: '10000000',
            },
            whitelistedAssets: {
              state: [
                {
                  getCode: jest.fn().mockReturnValue('EVRY'),
                  decimal: 1,
                },
              ],
            },
          })
          let updatedFormControls = { ...component.state().formControls }
          updatedFormControls.amount.value = '1'
          component.setState({
            formControls: updatedFormControls,
          })
          const form = component.find(Form)
          await form.props().onSubmit({
            preventDefault: () => {},
            target: {
              name: 'amount',
            },
          })
          expect(component).toMatchSnapshot()
        })
      })
    })
  })
})
