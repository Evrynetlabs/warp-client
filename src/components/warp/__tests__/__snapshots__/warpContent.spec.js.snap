// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WarpContent With default case should render correspondingly 1`] = `
<Form
  className="WarpContent__form"
  inline={false}
  onSubmit={[Function]}
>
  <CardBody
    className="WarpContent__form__content"
  >
    <Container
      fluid={true}
    >
      <ForwardRef
        noGutters={false}
      >
        <Col>
          <FormGroup
            controlId="sourceAccountNumber"
          >
            <FormLabel
              column={false}
              srOnly={false}
            >
              <span>
                From:
              </span>
            </FormLabel>
            <FormControl
              className="WarpContent__form__content__input__src WarpContent__form__content__input"
              onChange={[Function]}
              placeholder="Account Number"
              type="text"
              value=""
            />
          </FormGroup>
        </Col>
        <Col>
          <FormGroup
            controlId="destinationAccountNumber"
          >
            <FormLabel
              column={false}
              srOnly={false}
            >
              <span>
                To:
              </span>
            </FormLabel>
            <FormControl
              className="WarpContent__form__content__input__dest WarpContent__form__content__input"
              onChange={[Function]}
              placeholder="Account Number"
              type="text"
              value=""
            />
          </FormGroup>
        </Col>
      </ForwardRef>
      <ForwardRef
        className="WarpContent__form__content__amount"
        noGutters={false}
      >
        <Col>
          <FormGroup
            controlId="assetAmount"
          >
            <FormControl
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="0.00"
              type="text"
              value=""
            />
          </FormGroup>
        </Col>
        <Col>
          <FormGroup
            controlId="assetSelection"
          >
            <FormControl
              as="select"
              onChange={[Function]}
              value="FOO"
            />
          </FormGroup>
        </Col>
      </ForwardRef>
    </Container>
  </CardBody>
  <CardFooter
    className="WarpContent__footer"
  >
    <Container
      fluid={true}
    >
      <ForwardRef
        noGutters={false}
      >
        <Col
          className="WarpContent__footer__btn"
        >
          <Button
            active={false}
            disabled={false}
            type="submit"
            variant="dark"
          >
            Transfer
          </Button>
        </Col>
      </ForwardRef>
    </Container>
  </CardFooter>
</Form>
`;
